File: api/upscale.py
import base64
import io
import json
from PIL import Image
import numpy as np
import onnxruntime as ort


def handler(request):
    """
    Serverless handler para Vercel
    POST /api/upscale
    Body: {"image": "base64_string", "scale": 2 ou 4}
    Response: {"image": "base64_string_upscaled"}
    """
    if request.method != "POST":
        return {
            "statusCode": 405,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": "Method Not Allowed"})
        }

    try:
        body = request.get_json() if hasattr(request, 'get_json') else json.loads(request.body)
        image_base64 = body.get("image")
        scale = int(body.get("scale", 2))

        if not image_base64:
            return {
                "statusCode": 400,
                "headers": {"Content-Type": "application/json"},
                "body": json.dumps({"error": "No image provided"})
            }

        if scale not in [2, 4]:
            return {
                "statusCode": 400,
                "headers": {"Content-Type": "application/json"},
                "body": json.dumps({"error": "Scale must be 2 or 4"})
            }

        # Decodificar imagem base64
        image_bytes = base64.b64decode(image_base64)
        image = Image.open(io.BytesIO(image_bytes)).convert("RGB")

        # Preparar input para ONNX
        img_np = np.array(image).astype(np.float32) / 255.0
        img_np = np.transpose(img_np, (2, 0, 1))  # HWC -> CHW
        img_np = np.expand_dims(img_np, 0)  # Add batch dimension

        # Selecionar modelo baseado na escala
        model_path = f"python/RealESRGAN_x{scale}plus.onnx"

        # Executar inference com ONNX Runtime
        session = ort.InferenceSession(model_path, providers=["CPUExecutionProvider"])
        input_name = session.get_inputs()[0].name
        output = session.run(None, {input_name: img_np})[0]

        # Pós-processar output
        output = np.squeeze(output)
        output = np.transpose(output, (1, 2, 0))  # CHW -> HWC
        output = (output * 255.0).clip(0, 255).astype(np.uint8)
        out_image = Image.fromarray(output)

        # Retornar como base64
        buffer = io.BytesIO()
        out_image.save(buffer, format="PNG")
        img_b64 = base64.b64encode(buffer.getvalue()).decode()

        return {
            "statusCode": 200,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"image": img_b64})
        }

    except FileNotFoundError as e:
        return {
            "statusCode": 500,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": f"Modelo não encontrado: {str(e)}"})
        }
    except Exception as e:
        return {
            "statusCode": 500,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": str(e)})
        }

File: api/upscale.js
const express = require('express');
const bodyParser = require('body-parser');
const base64 = require('base64-js');
const { InferenceSession } = require('onnxruntime-node');
const Jimp = require('jimp');

const app = express();
app.use(bodyParser.json({ limit: '10mb' }));

app.post('/upscale', async (req, res) => {
  try {
    const { image } = req.body;
    if (!image) {
      return res.status(400).json({ error: 'No image provided' });
    }

    // Decodificar imagem base64
    const imageBuffer = Buffer.from(image, 'base64');
    const jimpImage = await Jimp.read(imageBuffer);
    jimpImage.background(0xFFFFFFFF);
    jimpImage.rgba(true);
    const { data, width, height } = jimpImage.bitmap;

    // Converter para Float32 e normalizar
    const imgArray = new Float32Array(width * height * 3);
    for (let i = 0; i < width * height; i++) {
      imgArray[i * 3 + 0] = data[i * 4 + 0] / 255.0;
      imgArray[i * 3 + 1] = data[i * 4 + 1] / 255.0;
      imgArray[i * 3 + 2] = data[i * 4 + 2] / 255.0;
    }

    // Formatar para [1, 3, H, W]
    const inputTensor = new Float32Array(1 * 3 * height * width);
    for (let c = 0; c < 3; c++) {
      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          inputTensor[c * height * width + y * width + x] = imgArray[y * width * 3 + x * 3 + c];
        }
      }
    }

    // Carregar modelo ONNX
    const session = await InferenceSession.create('python/model.onnx');
    const inputName = session.inputNames[0];
    const feeds = {};
    feeds[inputName] = inputTensor;
    const results = await session.run(feeds);
    const output = results[session.outputNames[0]];

    // Pós-processar saída para imagem
    const outHeight = output.dims[2];
    const outWidth = output.dims[3];
    const outData = output.data;
    const outImage = new Jimp(outWidth, outHeight);
    for (let y = 0; y < outHeight; y++) {
      for (let x = 0; x < outWidth; x++) {
        const r = Math.round(outData[0 * outHeight * outWidth + y * outWidth + x] * 255);
        const g = Math.round(outData[1 * outHeight * outWidth + y * outWidth + x] * 255);
        const b = Math.round(outData[2 * outHeight * outWidth + y * outWidth + x] * 255);
        outImage.setPixelColor(Jimp.rgbaToInt(r, g, b, 255), x, y);
      }
    }

    // Codificar imagem para base64
    const outBuffer = await outImage.getBufferAsync(Jimp.MIME_PNG);
    const outBase64 = outBuffer.toString('base64');
    res.json({ image: outBase64 });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

const PORT = process.env.PORT || 5000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});
